dags/synapse-by-the-numbers-dag.py

```diff
@@ -0,0 +1,159 @@
+"""This script is used to execute a query on Snowflake and report the results to a 
+Synapse table. This retrieves data describing the total amount of data hosted in Synapse,
+the number of active users last month, and the number of downloads last month."""
+
+from dataclasses import dataclass
+from datetime import date, datetime
+from typing import List
+
+import synapseclient
+from airflow.decorators import dag, task
+from airflow.models.param import Param
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+    "current_date": Param(date.today().strftime("%Y-%m-%d"), type="string"),
+}
+
+dag_config = {
+    # run on the 1st of every month at midnight
+    "schedule_interval": "0 0 1 * *",
+    "start_date": datetime(2024, 7, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+SYNAPSE_RESULTS_TABLE = "syn61588123"
+
+
+@dataclass
+class DownloadMetric:
+    """Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        total_data_size_in_pib: The size of the data hosted on Synapse in PiB
+        active_users_last_month: The number of active users last month
+        total_downloads_last_month: The total number of downloads by users last month
+    """
+
+    total_data_size_in_pib: float
+    active_users_last_month: int
+    total_downloads_last_month: int
+
+
+@dag(**dag_config)
+def synapse_by_the_numbers_past_month() -> None:
+    """Execute a query to gather the total amount of data hosted in Synapse,
+    the number of active users last month, and the number of downloads
+    last month from Snowflake and report the results to a Synapse table."""
+
+    @task
+    def get_synapse_monthly_metrics(**context) -> List[DownloadMetric]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+            WITH TOTAL_SIZE AS (
+                SELECT
+                    SUM(content_size) / POWER(2, 50) AS SIZE_IN_PETABYTES
+                FROM (
+                    SELECT DISTINCT
+                        file_latest.id,
+                        content_size
+                    FROM
+                        synapse_data_warehouse.synapse.node_latest
+                    JOIN
+                        synapse_data_warehouse.synapse.file_latest
+                        ON node_latest.file_handle_id = file_latest.id
+                    UNION
+                    SELECT DISTINCT
+                        file_latest.id,
+                        content_size
```

Comment:
Just a quick note: Are there duplicated file handle ids that is generated from this union?

Reply:
No:
https://docs.snowflake.com/en/sql-reference/operators-query#union-all

`UNION combines with duplicate elimination.`